File: src/main/java/com/qtech/bubbles/gui/QViewport.java
package com.qtech.bubbles.gui;

import com.qtech.bubbles.common.GraphicsProcessor;

import java.awt.*;

public class QViewport extends QContainer {
    private final Rectangle viewportRect;

    public QViewport(Rectangle viewportRect) {
        this.viewportRect = viewportRect;
        setBackgroundColor(new Color(64, 64, 64));
    }

    @Override
    public void renderComponents(GraphicsProcessor ngg) {
        for (QComponent component : components) {
            GraphicsProcessor componentGraphics = ngg.create(component.getX(), component.getY(), component.getWidth(), component.getHeight());
            component.render(componentGraphics);
            componentGraphics.dispose();
        }
    }

    @Override
    public void render(GraphicsProcessor ngg) {
        this.renderComponent(ngg);

        GraphicsProcessor viewportGraphics = ngg.create(viewportRect.x, viewportRect.y, viewportRect.width, viewportRect.height);
        renderComponents(viewportGraphics);
    }

    @Override
    public void renderComponent(GraphicsProcessor ngg) {
        ngg.setColor(getBackgroundColor());
        ngg.fillRect(0, 0, getSize().width, getSize().height);
    }

    @Override
    public void tick() {

    }

    public void setViewportSize(Dimension size) {
        this.viewportRect.setSize(size);
    }

    public Dimension getViewportSize() {
        return viewportRect.getSize();
    }

    public void setViewportLocation(Point location) {
        this.viewportRect.setLocation(location);
    }

    public Point getViewportLocation() {
        return viewportRect.getLocation();
    }
}

File: src/main/kotlin/qtech/bubbles/gui/QViewport.kt
package qtech.bubbles.gui

import qtech.hydro.GraphicsProcessor
import java.awt.Color
import java.awt.Dimension
import java.awt.Point
import java.awt.Rectangle

class QViewport(private val viewportRect: Rectangle) : QContainer() {
    override fun renderComponents(ngg: GraphicsProcessor) {
        for (component in components!!) {
            val componentGraphics = ngg.create(component.x, component.y, component.width, component.height)
            component.render(componentGraphics)
            componentGraphics.dispose()
        }
    }

    override fun render(ngg: GraphicsProcessor) {
        renderComponent(ngg)
        val viewportGraphics = ngg.create(viewportRect.x, viewportRect.y, viewportRect.width, viewportRect.height)
        renderComponents(viewportGraphics)
    }

    override fun renderComponent(ngg: GraphicsProcessor) {
        ngg.color(backgroundColor)
        ngg.rectF(0, 0, size.width, size.height)
    }

    override fun tick() {}
    var viewportSize: Dimension?
        get() = viewportRect.size
        set(size) {
            viewportRect.size = size
        }
    var viewportLocation: Point?
        get() = viewportRect.location
        set(location) {
            viewportRect.location = location
        }

    init {
        backgroundColor = Color(64, 64, 64)
    }
}